canonical-webteam/generator-canonical-webteam
# Add a basic run script for NPM-based projects

This builds on the run script in www.ubuntu.com and other projects.

It was initially added here by @nottrobin:
https://github.com/canonical-websites/www.ubuntu.com/pull/1411

And improved on by @WillMoggridge to add Yarn caching, namespacing of containers and detaching capability:
https://github.com/canonical-websites/www.ubuntu.com/pull/1459
https://github.com/canonical-websites/www.ubuntu.com/pull/1478

And further shores up the script with the following tweaks:

- Better encapsulating functions for running docker with specific
  settings
- Improve namespacing of containers so each container type has its own
  name - this prevents new `yarn` containers from killing existing
  containers except where appropriate (hopefully)
- Make `--detach` into an option on `./run start`, instead of a global
  option

QA
--

Install `yo`, link this branch and then create a dummy project directory with a package.json:

``` bash
sudo npm install -g yo
sudo npm link
cd ../
mkdir test-project
cd test-project
curl -s https://pastebin.com/raw/E2vy80Wx > package.json
```

Now run the generator:

``` bash
yo canonical-webteam      # See the help
yo canonical-webteam:run  # Install the ./run script
```

Test out the run script, check it makes sense:

``` bash
./run --help
./run
./run start --watch
./run start --detach
./run <other-things>
```
